components/prism-dot.js

```diff
@@ -0,0 +1,77 @@
+// https://www.graphviz.org/doc/info/lang.html
+
+(function (Prism) {
+
+	var ID = '(?:' + [
+		// an identifier
+		/\b[a-zA-Z_][\w\x80-\x9F\xA1-\xFF]*/.source,
+		// a number
+		/-?(?:\.\d+|\b\d+(?:\.\d*)?)/.source,
+		// a double-quoted string
+		/"(?:[^\\"]|\\[\s\S])*"/.source,
+		// HTML-like string
+		/<(?:[^<>]|(?!<!--)<(?:[^<>"']|"[^"]*"|'[^']*')+>|<!--(?:[^-]|-(?!->))*-->)*>/.source
+	].join('|') + ')';
+
+	var IDInside = {
+		'markup': {
+			pattern: /(^<)[\s\S]+(?=>$)/,
+			lookbehind: true,
+			alias: ['language-markup', 'language-html', 'language-xml'],
+			inside: Prism.languages.markup
+		}
+	};
+
+	/**
+	 * @param {string} source
+	 * @param {string} flags
+	 * @returns {RegExp}
+	 */
+	function withID(source, flags) {
+		return RegExp(source.replace(/<ID>/g, function () { return ID; }), flags);
+	}
+
+	Prism.languages.dot = {
+		'comment': {
+			pattern: /\/\/.*|\/\*[\s\S]*?\*\/|^#.*/m,
```

Comment:
How about changing `\/\/` to `\/{2}`? The latter takes 1 step less.

Reply:
`\/\*[\s\S]*?\*\/` backtracks on every character inside a comment.

|Comment       |Steps|Backtracks|
|--------------|-----|----------|
|/\*\*/          |    5|         0|
|/\*1\*/         |    7|         1|
|/\*\*\*/         |    8|         1|
|/\*12\*/        |    9|         2|
|/\*\*\*\*/        |   11|         2|
|/\*123\*/       |   11|         3|
|/\*\*\*\*\*/       |   14|         3|
|/\*1234567890\*/|   25|        10|
|/\*\*\*\*\*\*\*\*\*\*\*\*/|   35|        10|

+3 steps for every asterisk and +2 steps for any other character.

As I see, the amount of steps/backtracks grows linearly. But anyway, I hope, and please correct me if I’m wrong, that

```
\/\*[^*]*\*+(?:[^*/][^*]*\*+)*\/
```

should always make a single backtrack. In most cases the pattern takes **7 steps**. The amount of steps grows linearly only on repeating “an asterisk – not an asterisk” substrings.

|Comment     |Steps|Backtracks|
|------------|-----|----------|
|/\*\*/      |    7|         1|
|/\*1\*/     |    7|         1|
|/\*\*\*/    |    7|         1|
|/\*12\*/    |    7|         1|
|...         |  ...|       ...|
|/\*\*1\*/   |   10|         1|
|/\*\*1\*\*/ |   10|         1|
|/\*\*1\*1\*/|   14|         1|

Reply:
> How about changing `\/\/` to `\/{2}`? The latter takes 1 step less.

Most regex engines actually desugar `[char]{x}` into `[char][char]...[char]` (x-many).

I don't how what tool you use to measure these steps but its underlying regex engine model may not be one used in browsers.

> `\/\*[\s\S]*?\*\/` backtracks on every character inside a comment.

Please be careful! Steps and backtracks are not the same as actual runtime. For those two patterns in particular, the simpler, seemingly-less-efficient one is actually faster.

<details>

```js
measure = fn => { let s = performance.now(); fn(); return performance.now() - s; };

var r1 = /\/\*[\s\S]*?\*\//;
var r2 = /\/\*[^*]*\*+(?:[^*/][^*]*\*+)*\//;

// nodejs repl
> measure(() => r1.test( '/*' + "*".repeat(1e6) ))
3.105300188064575
> measure(() => r2.test( '/*' + "*".repeat(1e6) ))
7.263500213623047

> measure(() => r1.test( '/*' + "a".repeat(1e6) ))
2.6819000244140625
> measure(() => r2.test( '/*' + "a".repeat(1e6) ))
4.4449992179870605

> measure(() => r1.test( '/*' + "*".repeat(1e6) + '*/' ))
1.992799997329712
> measure(() => r2.test( '/*' + "*".repeat(1e6) + '*/' ))
2.0360000133514404

> measure(() => r1.test( '/*' + "a".repeat(1e6) + '*/' ))
1.9930000305175781
> measure(() => r2.test( '/*' + "a".repeat(1e6) + '*/' ))
1.8396999835968018

> measure(() => r1.test( '/*' + "a*".repeat(1e6) ))
4.826600074768066
> measure(() => r2.test( '/*' + "a*".repeat(1e6) ))
49.58860111236572

> measure(() => r1.test( '/*' + "a*".repeat(1e6) + '*/' ))
3.439500093460083
> measure(() => r2.test( '/*' + "a*".repeat(1e6) + '*/' ))
44.13789987564087
```

</details>

I know. _Why is the less efficient one faster?_ My best guess is that the regex engine is just better at optimizing simple patterns.

Reply:
> I don't how what tool you use to measure these steps

I use RegexBuddy (a trial version).

> > measure(() => r1.test( '/*' + "a*".repeat(1e6) ))
> 4.826600074768066
> > measure(() => r2.test( '/*' + "a*".repeat(1e6) ))
> 49.58860111236572

Wow.

> the regex engine is just better at optimizing simple patterns.

Ah, thank you very much!